src/java.logging/share/classes/java/util/logging/Handler.java

```diff
@@ -73,6 +76,17 @@ public abstract class Handler {
      * as the {@code ErrorManager}.
      */
     protected Handler() {
+        lock = initLocking();
+    }
+
+    private InternalLock initLocking() {
+        Class<?> clazz = this.getClass();
+        ClassLoader loader = clazz.getClassLoader();
+        if (loader != null && loader != ClassLoader.getPlatformClassLoader()) {
+            return null;
+        } else {
+            return InternalLock.newLockOrNull();
```

Comment:
I can revert to using plain `ReentrantLock` if you think it's preferable.

Reply:
> It's the same reason here: in these classes (and before that change) the lock is `this` which is always exposed to subclasses or external classes. If a handler uses `InternalLock`, and an external class `synchronize(handler)` that could cause surprising effects. My first take at this was simply using `new ReantrantLock()` but I thought it made sense to reuse `InternalLock` instead. After all, there would be no point in not using `synchronized` in StreamHandler if the underlying output stream is a PrintStream for which use of InternalLock has been disabled?

The reason for InternalLock is because the Reader/Write "lock" field is exposed to subclasses and there is a possibility that a subclass could set the lock field to an instance of ReentrantLock and confusing all the locking.  You don't have this issue in j.u.logging. I am not objecting to using InternalLock, just surprised to see it being used here as I had assumed you'd just create your own explicit lock when not subclassed.

Reply:
It's the same usage than in `PrintStream`: the lock in `PrintStream` is an `InternalLock` even though it's never exposed to subclasses (it's a private field). My rationale was that if the underlying PrintStream uses synchronized and doesn't use `InternalLock`, because `-Djdk.io.useMonitors=true`, then there's no point in the `Handler` trying to avoid using synchronized. Though I admit that not all `Handlers` wrap a `PrintStream`, the `FileHandler` and `ConsoleHandler` (which are the more important ready-to-use concrete implementations) will eventually delegate to some underlying IO class that will be impacted by `-Djdk.io.useMonitors=true`. So I was thinking that we could/should use the same logic there.

Reply:
Ah - I see that `PrintStream` lock can be accessed through SharedSecrets... Hmmm. OK - then maybe I should leave InternalLock alone and just use ReentrantLock. Let me prototype that.

Reply:
> Ah - I see that PrintStream lock can be accessed through SharedSecrets

I think Alan was talking about the 
```java
protected Object lock
```
on `java.io.Reader` and `java.io.Writer` classes which can be manipulated by their sub-classes and that's what prompted the creation of the new type, - `InternalLock` class.

Reply:
Yes - but InternalLock is not only used in those cases.